File: solutions/633. Sum of Square Numbers.py
class Solution:
    def judgeSquareSum(self, c: int) -> bool:
        ### Soln - native attempt, look at sqrt(c - i*i) for i = 0,1,2,...
        n = floor(sqrt( 2**31 - 1))
        for i in range(n):
            i2 = i*i
            if i2 > c:
                break
            else:
                m = sqrt(c - i*i)
                if m == floor(m):
                    return True
        return False
        

File: solutions/633. Sum of Square Numbers.cpp
class Solution {
public:
    bool judgeSquareSum(int c) {        
        // Brute-force
        for (int x = 0; x <= floor(sqrt(c/2)); x++) {
            int d = c - x*x;
            int dSqrt = floor(sqrt(d));
            if (dSqrt*dSqrt == d) return true;
        }
        return false;
        
        // // Fermat theorem
        // for (int x = 2; x*x <= c; x++) {
        //     if (c % x == 0) {
        //         int count = 0;
        //         while (c % x == 0) {
        //             count++;
        //             c /= x;
        //         }
        //         if (count % 2 != 0 && x % 4 == 3) return false;
        //     }
        // }
        // return c % 4 != 3;
    }
};
